src/dune/action_exec.ml

```diff
@@ -1,51 +1,90 @@
 open! Stdune
 open Import
 open Fiber.O
+module DAP = Dune_action_plugin.Private.Protocol
 
-module Context = struct
-  module For_exec = struct
+module Dynamic_dep = struct
+  module T = struct
     type t =
-      { working_dir : Path.t
-      ; env : Env.t
-      ; stdout_to : Process.Io.output Process.Io.t
-      ; stderr_to : Process.Io.output Process.Io.t
-      ; stdin_from : Process.Io.input Process.Io.t
-      }
+      | File of Path.t
+      | Glob of Path.t * Glob.t
+
+    let to_dep = function
+      | File fn -> Dep.file fn
+      | Glob (dir, glob) ->
+        Glob.to_pred glob |> File_selector.create ~dir |> Dep.file_selector
+
+    let of_DAP_dep ~working_dir : DAP.Dependency.t -> t =
+      let to_dune_path = Stdune.Path.relative working_dir in
+      function
+      | File fn -> File (to_dune_path fn)
+      | Directory dir -> Glob (to_dune_path dir, Glob.universal)
+
+    let compare x y =
+      match (x, y) with
+      | File x, File y -> Path.compare x y
+      | File _, _ -> Lt
+      | _, File _ -> Gt
+      | Glob (dir1, glob1), Glob (dir2, glob2) ->
+        Tuple.T2.compare Path.compare Glob.compare (dir1, glob1) (dir2, glob2)
+
+    let to_dyn =
+      let open Dyn.Encoder in
+      function
+      | File fn -> constr "File" [ Path.to_dyn fn ]
+      | Glob (dir, glob) -> constr "Glob" [ Path.to_dyn dir; Glob.to_dyn glob ]
   end
 
-  type t =
-    { context : Context.t option
-    ; purpose : Process.purpose
-    ; initial_env : Env.t
-    }
+  include T
+  module O = Comparable.Make (T)
 
-  let env t = t.initial_env
+  module Set = struct
+    include O.Set
 
-  let for_exec t =
-    { For_exec.working_dir = Path.root
-    ; env = t.initial_env
-    ; stdout_to = Process.Io.stdout
-    ; stderr_to = Process.Io.stderr
-    ; stdin_from = Process.Io.stdin
-    }
+    let to_dep_set t = t |> to_list |> List.map ~f:to_dep |> Dep.Set.of_list
 
-  let make ~targets ~(context : Context.t option) ~env =
-    let initial_env =
-      match (env, context) with
-      | None, None -> Env.initial
-      | Some e, _ -> e
-      | None, Some c -> c.env
-    in
-    let purpose = Process.Build_job targets in
-    { purpose; context; initial_env }
+    let of_DAP_dep_set ~working_dir t =
+      t |> DAP.Dependency.Set.to_list
+      |> List.map ~f:(of_DAP_dep ~working_dir)
+      |> of_list
+  end
+end
+
+module Exec_result = struct
+  type t = { dynamic_deps_stages : Dynamic_dep.Set.t List.t }
 end
 
-let exec_run ~(ectx : Context.t) ~(eenv : Context.For_exec.t) prog args =
-  ( match ectx.context with
+type done_or_more_deps =
+  | Done
+  (* This code assumes that there can be at most one 'dynamic-run' within
+     single action. [DAP.Dependency.t] stores relative paths so name clash
+     would possible if multiple 'dynamic-run' would be executed in different
+     subdirectories that contains targets having the same name. *)
```

Comment:
Aha, I found an explanation here. It's a bit sad to rule out legitimate uses of multiple `dynamic-run`s because of possible clashes on targets with the same name. 

Also "would possible if multiple" -> "would be possible if multiple".

Reply:
I don't find this to be particularly sad, but maybe I'm missing something. I think we could support it pretty easily if we really had a use case.